apis/refs/v1beta1/computerefs.go

```diff
@@ -269,23 +365,120 @@ type ComputeTargetTCPProxyRef struct {
+	}
+
+	if ref.Name == "" {
+		return nil, fmt.Errorf("must specify either name or external on ComputeTargetVPNGateway reference")
+	}
+
+	key := types.NamespacedName{
+		Namespace: ref.Namespace,
+		Name:      ref.Name,
+	}
+	if key.Namespace == "" {
+		key.Namespace = src.GetNamespace()
+	}
+
+	computeTargetVPNGateway := &unstructured.Unstructured{}
+	computeTargetVPNGateway.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   "compute.cnrm.cloud.google.com",
+		Version: "v1beta1",
+		Kind:    "ComputeTargetVPNGateway",
+	})
+	if err := reader.Get(ctx, key, computeTargetVPNGateway); err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil, fmt.Errorf("referenced ComputeAddress %v not found", key)
+		}
+		return nil, fmt.Errorf("error reading referenced ComputeAddress %v: %w", key, err)
+	}
+
+	computeTargetVPNGatewayID, err := GetResourceID(computeTargetVPNGateway)
+	if err != nil {
+		return nil, err
+	}
+
+	computeTargetVPNGatewayProjectID, err := ResolveProjectID(ctx, reader, computeTargetVPNGateway)
+	if err != nil {
+		return nil, err
+	}
+
+	computeTargetVPNGatewayLocation, err := getLocation(computeTargetVPNGateway)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ComputeTargetVPNGateway{
+		Project:                   computeTargetVPNGatewayProjectID,
+		Location:                  computeTargetVPNGatewayLocation,
+		ComputeTargetVPNGatewayID: computeTargetVPNGatewayID,
+	}, nil
+}
+
 // TODO(yuhou): Location can be optional. Use provider default location when it's unset.
 func getLocation(obj *unstructured.Unstructured) (string, error) {
-	// TODO(yuhou): field can be "location" or "region".
-	location, _, err := unstructured.NestedString(obj.Object, "spec", "location")
+	location, found, err := unstructured.NestedString(obj.Object, "spec", "location")
 	if err != nil {
 		return "", fmt.Errorf("cannot get location for referenced %s %v: %w", obj.GetKind(), obj.GetNamespace(), err)
 	}
+	if !found {
+		// if region is set, use its value as location
+		location, found, err = unstructured.NestedString(obj.Object, "spec", "region")
+		if err != nil {
+			return "", fmt.Errorf("cannot get region for referenced %s %v: %w", obj.GetKind(), obj.GetNamespace(), err)
+		}
+		if !found {
+			return "", fmt.Errorf("cannot get location or region for referenced %s %v (spec.location or spec.region not set)", obj.GetKind(), obj.GetNamespace())
+		}
+	}
 	if location == "" {
-		return "", fmt.Errorf("cannot get location for referenced %s %v (spec.location not set)", obj.GetKind(), obj.GetNamespace())
+		return "", fmt.Errorf("cannot get location or region for referenced %s %v (spec.location or spec.region is empty)", obj.GetKind(), obj.GetNamespace())
 	}
 	return location, nil
 }
+
+func getAddress(obj *unstructured.Unstructured) (string, error) {
+	address, _, err := unstructured.NestedString(obj.Object, "spec", "address")
```

Comment:
Nit: can we add a TODO that we should be source this from status or status.observedState?